File: system-test/java/src/main/java/com/optivem/eshop/dsl/core/usecase/external/erp/ErpDsl.java
package com.optivem.eshop.dsl.core.usecase.external.erp;

import com.optivem.eshop.dsl.driver.port.external.erp.ErpDriver;
import com.optivem.eshop.dsl.core.usecase.external.erp.usecases.GetProduct;
import com.optivem.eshop.dsl.core.usecase.external.erp.usecases.GoToErp;
import com.optivem.eshop.dsl.core.usecase.external.erp.usecases.ReturnsProduct;
import com.optivem.eshop.dsl.common.Closer;
import com.optivem.eshop.dsl.core.shared.UseCaseContext;

public class ErpDsl implements AutoCloseable {
    protected final ErpDriver driver;
    protected final UseCaseContext context;

    public ErpDsl(ErpDriver driver, UseCaseContext context) {
        this.driver = driver;
        this.context = context;
    }

    @Override
    public void close() {
        Closer.close(driver);
    }

    public GoToErp goToErp() {
        return new GoToErp(driver, context);
    }

    public ReturnsProduct returnsProduct() {
        return new ReturnsProduct(driver, context);
    }

    public GetProduct getProduct() {
        return new GetProduct(driver, context);
    }
}

File: system-test/typescript/src/testkit/dsl/core/usecase/external/erp/ErpDsl.ts
import type { Result } from '../../../../../common/result.js';
import type { ErpDriver } from '../../../../../driver/port/external/erp/erp-driver.js';
import type { ErpErrorResponse } from '../../../../../driver/port/external/erp/dtos/ErpErrorResponse.js';
import type { UseCaseContext } from '../../../use-case-context.js';
import { ReturnsProduct } from './usecases/ReturnsProduct.js';

export class ErpDsl {
  constructor(
    private readonly driver: ErpDriver,
    private readonly context: UseCaseContext,
  ) {}

  async goToErp(): Promise<Result<void, ErpErrorResponse>> {
    return this.driver.goToErp();
  }

  returnsProduct(): ReturnsProduct {
    return new ReturnsProduct(this.driver, this.context);
  }

  async close(): Promise<void> {
    await this.driver.close();
  }
}
